libecole/include/ecole/environment/environment.hpp

```diff
@@ -112,22 +112,30 @@ class Environment {
 			dynamics().set_dynamics_random_state(model(), random_engine());
 
 			// Reset data extraction function and bring model to initial state.
-			observation_function().before_reset(model());
 			reward_function().before_reset(model());
+			observation_function().before_reset(model());
 			information_function().before_reset(model());
-			auto const [done, action_set] = dynamics().reset_dynamics(model(), std::forward<Args>(args)...);
+
+			// Place the environment in its initial state
+			auto&& [done, action_set] = dynamics().reset_dynamics(model(), std::forward<Args>(args)...);
 			can_transition = !done;
 
-			auto observation = OptionalObservation{};
-			if (!done) {
-				observation = observation_function().extract(model(), done);
-			}
+			// Extract reward, observation and information (in that order)
+			auto&& reward = reward_function().extract(model(), done);
+			auto&& observation = [&, done = done]() -> OptionalObservation {
+				if (done) {  // Don't extract observations in final states
+					return {};
+				}
+				return observation_function().extract(model(), done);
+			}();
+			auto&& information = information_function().extract(model(), done);
```

Comment:
```suggestion
			auto information = information_function().extract(model(), done);
```

Reply:
This is true for gcc, but for clang `done=done` is required, otherwise it finds our code is not valid. Something to do with `done` being a structured binding. See:
https://stackoverflow.com/questions/46114214/lambda-implicit-capture-fails-with-variable-declared-from-structured-binding